File: server/lib/channels/store.ts
import { parseM3U } from './m3u'
import type { Channel, ChannelLogger } from './types'

const noopLogger: ChannelLogger = {
  warn() {}
}

export class ChannelStore {
  #channels: Channel[] = []
  #logger: ChannelLogger

  constructor(logger: ChannelLogger = noopLogger) {
    this.#logger = logger
  }

  replaceFromRaw(playlist: string): Channel[] {
    const nextChannels = parseM3U(playlist, { logger: this.#logger })
    this.#channels = nextChannels
    return this.getChannels()
  }

  getChannels(): Channel[] {
    return [...this.#channels]
  }
}

File: internal/channels/store.go
package channels

import (
	"errors"
	"sync"
)

type Store struct {
	channels []Channel
	logger   WarnLogger
	mu       sync.RWMutex
}

func NewStore(logger WarnLogger) *Store {
	return &Store{logger: logger}
}

func (s *Store) ReplaceFromRaw(playlist string) []Channel {
	next := ParseM3U(playlist, s.logger)
	s.mu.Lock()
	s.channels = next
	s.mu.Unlock()
	return s.GetChannels()
}

func (s *Store) Refresh(fetchPlaylist func() (string, error)) ([]Channel, error) {
	playlist, err := fetchPlaylist()
	if err != nil {
		return nil, err
	}
	next := ParseM3U(playlist, s.logger)
	if len(next) == 0 {
		return nil, errors.New("zero valid channels")
	}
	s.mu.Lock()
	s.channels = next
	s.mu.Unlock()
	return s.GetChannels(), nil
}

func (s *Store) GetChannels() []Channel {
	s.mu.RLock()
	defer s.mu.RUnlock()
	next := make([]Channel, len(s.channels))
	copy(next, s.channels)
	return next
}
